### ISSUE TITLE
Omiting any file/dir positional parameter to CLI does not give an error

### ISSUE THREAD

**Issue by jobedom-stylesage** [eslint/eslint#14308]
* **ESLint Version:** 7.23.0
* **Node Version:** 15.4.0
* **npm Version:** 7.1.1
* **Operating System:** linux 5.4.0-70-generic
* **Parser:** @babel/eslint-parser

<details>
<summary>Configuration</summary>

```js
module.exports = {
  root: true,
  parserOptions: {
    parser: '@babel/eslint-parser',
    ecmaVersion: 2018,
    sourceType: 'module'
  },
  extends: [
    'standard',
    'plugin:vue/recommended'
  ],
  plugins: ['vue'],
  rules: {
    'max-len': ['warn', { code: 120 }],
    'vue/no-v-html': 'off'
  }
}
```

</details>

ESLint CLI usage:

```bash
eslint --cache --ext .js,.vue --max-warnings 0
```

**What did you expect to happen?**

I expected to have and error in console: `'patterns' must be a non-empty string or an array of non-empty strings`.

Instead, I got no response at all (and no actual linting performed, of course). 

```
$ eslint --cache --ext .js,.vue --max-warnings 0
Done in 0.23s.
```

**How to replicate?**

1. Run `eslint` CLI with no positional parameters. 
2. No error is shown .

**More details**

The [condition in ESLint engine code](https://github.com/eslint/eslint/blob/71a80e38aab2dada01b808ed43d9b0e806d863c4/lib/eslint/eslint.js#L523) is accepting an empty list as a valid list of non empty strings. It boils down to the fact that a call to `.every(...)` method of an empty list is always true ([see `isArrayOfNonEmptyString` function](https://github.com/eslint/eslint/blob/71a80e38aab2dada01b808ed43d9b0e806d863c4/lib/eslint/eslint.js#L119))

I'm more than OK to prepare a PR to fix this.

Thanks!

**Comment by nzakas** [comment]
We will need a minimal reproduction case in order to look at this further. Without knowing which files are being used and what warnings might be in them, there’s not enough info to determine anything.

Is this in a public repo? If not, can you create a minimal repro case in a public repo so we can see the problem in action?

**Comment by jobedom-stylesage** [comment]
I can create a minimal repo if it helps, but my report refers to a missing error from ESLint, not a missing linting error per se, so it applies to any repo using ESLint.

I'll try to explain: I think running the CLI without any positional parameter (no dir or file whatsoever), in any installation of ESLint, should give the ESLint (not linting) error `'patterns' must be a non-empty string or an array of non-empty strings` and it does not, as explained in the issue.

Let me know if something else is needed.

**Comment by nzakas** [comment]
Ah, I see what you mean. I couldn’t quite figure out what you meant by “positional parameter” but now I see that you mean no glob patterns were passed on the command line. Thanks for explaining.

yes, this seems like a bug. We would appreciate a PR if you’re up for it.

**Comment by nzakas** [comment]
I will do this as part of new config system work.

**Comment by drzraf** [comment]
Something apparently made its way to the codebase at some point and is annoying me.

I used to have the following in my `package.json`:
`lint:js:staged": "eslint $(git diff --cached --name-only --relative -- '*.js') ,`

Then, to avoid an error when no JS file has been changed, I added an always present empty argument:
`lint:js:staged": "eslint $(git diff --cached --name-only --relative -- '*.js') ''",`

But this trick doesn't work anymore.
How to make eslint exit smoothly if the file selection is empty, now?

**Comment by nzakas** [comment]
@drzraf please start a discussion or stop by Discord for this type of question.

**Comment by drzraf** [comment]
Aside from discussing on Discord, I think it's an actual bug (or rather disappeared feature). Don't you think the former behavior were failing safe was privileged over argument validation/consistency would be desirable.
eslint being a tool used within editors, command-line & CI, I do think it has great advantage to handle the absence of file arguments.

**Comment by mdjermanovic** [comment]
Should we throw an error, or use the config file to determine which files to lint - per the [RFC](https://github.com/eslint/rfcs/blob/main/designs/2019-config-simplification/README.md):

> Additionally, ESLint can be run without specifying anything on the command line, relying just on what's in `eslint.config.js` to determine what to lint:
> 
> ```bash
> eslint
> ```
>
> This will go into the `eslint.config.js` and use all of the `files` glob patterns to determine which files to lint.

I believe the latter would mean that when no positional arguments (file/dir/glob lint targets) are specified, the default is `["."]`.

**Comment by drzraf** [comment]
Sound fine to handle omission or file arguments by use of a default argument **if** there were a way to "fail-safe" **too**. Eg: If the ` ""`  empty argument (or a `--`  or a non-existing file) would not trigger an error.

**Comment by nzakas** [comment]
> I believe the latter would mean that when no positional arguments (file/dir/glob lint targets) are specified, the default is ["."].

Not exactly. It would mean that ESLint wouldn't assuming anything and instead take all of the `files` patterns from `eslint.config.js` and use those as the glob patterns. So `files:["**/*.js"]` plus `eslint` equals `eslint **/*.js`.

**Comment by MaximeBernard** [comment]
> I believe the latter would mean that when no positional arguments (file/dir/glob lint targets) are specified, the default is `["."]`.

Agree with that. Feels quite intuitive 👍 

Just happened that one of our packages was successfully linting a package for years because we missed the `.` in `eslint .`. 

`eslint` should either 
- lint `.` (or `files` parameter in eslint config)
- exit with an error `no parameter passed`

Current behavior is misleading.

**Comment by nzakas** [comment]
I'm starting work on this. In eslintrc mode, `eslint` will now error; in flat config mode, `eslint` will now be equivalent to `eslint .`.


In order to have a way to opt back in to the current behavior, where you can pass zero file patterns to ESLint and it won't error, we're thinking to add a `--no-error-on-no-patterns` flag. Thoughts?

**Comment by mdjermanovic** [comment]
For eslintrc mode, where passing zero patterns will be error, `--no-error-on-no-patterns` to suppress the error sounds good to me.

For flat config mode, where passing zero patterns will mean "lint everything you find", I think we'd need a different name for an option that prevents any linting and just exits with zero. Maybe something like `--no-linting-on-no-patterns` or `--exit-on-no-patterns` (though we already have option `--exit-on-fatal-error` with a similar name but very different impact on the exit code so that might be confusing).

**Comment by nzakas** [comment]
I kind of like `--exit-on-no-patterns`. Or maybe something like `--pass-on-no-patterns`? Or `--ok-on-no-patterns`?

**Comment by nzakas** [comment]
Ping @mdjermanovic

**Comment by mdjermanovic** [comment]
`--pass-on-no-patterns` sounds good to me for both eslintrc and flat config modes.

**Comment by sam3k** [comment]
Per 2023-11-30 tsc-meeting; this issue is ready for review.
